nitransforms/linear.py

```diff
@@ -35,34 +42,29 @@ def __init__(self, matrix=None, reference=None):
         --------
         >>> xfm = Affine(reference=datadir / 'someones_anatomy.nii.gz')
         >>> xfm.matrix  # doctest: +NORMALIZE_WHITESPACE
-        array([[[1., 0., 0., 0.],
-                [0., 1., 0., 0.],
-                [0., 0., 1., 0.],
-                [0., 0., 0., 1.]]])
+        array([[1., 0., 0., 0.],
+               [0., 1., 0., 0.],
+               [0., 0., 1., 0.],
+               [0., 0., 0., 1.]])
 
         >>> xfm = Affine([[1, 0, 0, 4], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]])
         >>> xfm.matrix  # doctest: +NORMALIZE_WHITESPACE
-        array([[[1, 0, 0, 4],
-                [0, 1, 0, 0],
-                [0, 0, 1, 0],
-                [0, 0, 0, 1]]])
+        array([[1, 0, 0, 4],
+               [0, 1, 0, 0],
+               [0, 0, 1, 0],
+               [0, 0, 0, 1]])
 
         """
-        super(Affine, self).__init__()
-        if matrix is None:
-            matrix = [np.eye(4)]
-
-        if np.ndim(matrix) == 2:
-            matrix = [matrix]
-
-        self._matrix = np.array(matrix)
-        assert self._matrix.ndim == 3, 'affine matrix should be 3D'
-        assert (
-            self._matrix.shape[-2] == self._matrix.shape[-1]
-        ), 'affine matrix is not square'
+        super().__init__(reference=reference)
+        self._matrix = np.eye(4)
 
-        if reference:
-            self.reference = reference
+        if matrix is not None:
+            matrix = np.array(matrix)
+            if matrix.ndim != 2:
+                raise TypeError('Affine should be 2D.')
```

Comment:
Since this matrix is the core formal element and the problem is not whether the values in the matrix make sense or are from a valid finding of inputs, this seemed to me more like a type error. Let me know if this argument is convincing

Reply:
At least to me, it doesn't make sense to throw a `TypeError` if the matrix is an array, but it does make sense to throw a `ValueError` if the array dimensions aren't what is expected. But your call at end, nbd

Reply:
Okay, this is my last attempt to show what I think is a `ValueError` in contrast to a `TypeError`: https://github.com/poldracklab/nitransforms/pull/59/files#diff-ca10ddac142c49ab69b356fb51a6a075R69-R70

Here, all the formal checks (i.e., shape) on the validity of the matrix pass, but the values in it are outside of the possible inputs domain.

The two type errors above could go away if we modify this object to, for instance, take a (3, 4) matrix (à la AFNI), or accept a mat+vec representation (allow the argument matrix to be (3, 3) and have an extra (3, ) vector argument for the translation).

The ValueError,  though, will always be there.

This is completely a conceptual and possibly arbitrary argument - if you are still unconvinced I'll assume I'm wrong (that's the most likely possibility in that case because you are an English speaker and I'm not) and switch all to `ValueError`s.